enhancements/baremetal-config.md

```diff
@@ -0,0 +1,153 @@
+
+### Proposal
+
+To support configuration of a Baremetal IPI (metal3) deployment, a set of new
+config items are required to be made availabe via a Config Resource. The
+BareMetalPlatformStatus is an ideal fit because it contains configuration
+that is specific to a Baremetal deployment.
+
+This proposal aims to add the following to BareMetalPlatformStatus:
+
+1. ProvisioningInterface : This is the interface name on the underlying
+baremetal server which would be physically connected to the provisioning
+network.
+
+2. ProvisioningIP : This is the IP address used to to bring up a NIC on
+the baremetal server for provisioning. This value should not be in the
+DHCP range and should not be used in the provisioning network for any
+other purpose (should be a free IP address.) It is expected to be provided
+in the format : 10.1.0.3.
+
+3. ProvisioningNetworkCIDR : This is the provisioning network with its
+CIDR. The ProvisioningIP and the ProvisioningDHCPRange are expected to
+be within this network. The value for this config is expected to be
+provided in the format : 10.1.0.0/24.
+
+4. ProvisioningDHCPRange : This is a range of IP addresses from which
+baremetal hosts can be assigned their IP address. This range should be
+provided as two comma seperated IP addresses where the addresses represent
+a contiguous range of IPs that are available to be allocated. It is expected
+to be provided in the format : 10.1.0.10, 10.1.0.100.
+
+5. CacheURL : This is a location where image required to bring up baremetal
+hosts have been previously downloaded for faster image downloads within the
+cluster. This is an optional parameter and is expected to be provided
+in the format : http://111.111.111.1/images
+
+6. RHCOSImageURL : This config is expected to be the location of the
+official RHCOS release image. It is a required parameter and will be used
+to deploy new nodes in the cluster.
+
+### User Stories [optional]
+
+This enhancement will allow users to have their Barametal IPI deployments
+customizable to their specific hardware and network requirements.
+
+This also allows IPI baremetal deployments to take place without manual
+workarounds like creating a ConfigMap for the config (which is the
+current approach we are using to test baremetal deployments.) 
+
+The new config items would be set by the installer and will be used by
+the machine-api-operator (MAO) to generate more config items that are
+derivable from these basic parameters. Put together, these config
+parameters are passed in as environment variables to the various containers
+that make up a metal3 baremetal IPI deployment. 
+
+## Design Details
+
+### Test Plan
+
+The test plan should involve making sure the openshift/installer generates
+all configuration items within the BaremetalPlatformStatus when the platform
+type is Baremetal.
+
+MAO reads this configuration and uses these to derive additional configuration
+required to bring up a metal3 cluster. E2e testing should make sure that MAO
```

Comment:
What type of negative testing do you plan on doing?  Malformed strings, IP addresses/ranges overlapping, not in the expected ranges, poorly formed IP addresses, images not where they are specified, slow or interrupted downloads.  Is this going to be tested and within scope of this enhancement testing?

Reply:
> What type of negative testing do you plan on doing? Malformed strings, IP addresses/ranges overlapping, not in the expected ranges, poorly formed IP addresses, images not where they are specified, slow or interrupted downloads. Is this going to be tested and within scope of this enhancement testing?

All of these values will be provided by the installer ref openshift/installer#2449 - that does have unit tests for the installer level validation of those values, and I'd assume we'll have similar validation and unit tests in the MAO @sadasu ?

For e2e testing we've got a 3rd party metal3 CI job which can be used to validate these changes, but not yet fully integrated openshift e2e CI for this new platform - @derekhiggins is working on that.

Reply:
Also note that the only way to use this platform at present is to manually create a ConfigMap resource, e.g a completely unconstrained interface, so IMO having the validation at the installer layer will be a big improvement.